velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +782,133 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+    size_t width) {
+  for (auto j = 0; j < width; ++j) {
+    if (currentRow[j] != previousRow[j]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Return true if there are multiple rows with the same prefix of the first
+// width columns. The argument groupedBuckets should be sorted by the first
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& groupedBuckets,
+    size_t width) {
+  for (auto i = 1; i < groupedBuckets.size(); ++i) {
+    if (equalPrefix(groupedBuckets[i], groupedBuckets[i - 1], width)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Try to bucket rows in left and right into leftBucketed and rightBucketed
+// and returns whether the bucketization is successful. The bucketization is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
+//   same types.
+//   2. Rows in left and right are grouped by non-floating-point column. If
+//   there are multiple floating-point columns, either left or right must have a
+//   single row in every group. This is not required in case of zero or one
+//   floating-point column.
+// The bucket results are stored in leftBucketed and rightBucketed with the
+// floating-point columns moved to the end and rows in ascending order.
+bool tryBucketMaterializedRows(
+    const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right,
+    std::vector<MaterializedRow>& leftBucketed,
+    std::vector<MaterializedRow>& rightBucketed) {
+  if (left.empty() || right.empty() ||
+      left.begin()->size() != right.begin()->size()) {
+    return false;
+  }
+
+  std::vector<velox::column_index_t> floatColumnIndices;
+  auto& leftRow = *left.begin();
+  auto& rightRow = *right.begin();
+  auto width = leftRow.size();
+  for (auto i = 0; i < width; ++i) {
+    if (leftRow[i].kind() != rightRow[i].kind()) {
+      return false;
+    }
+    if ((leftRow[i].kind() == TypeKind::REAL ||
+         leftRow[i].kind() == TypeKind::DOUBLE)) {
+      floatColumnIndices.push_back(i);
+    }
+  }
+  if (floatColumnIndices.empty()) {
+    return false;
+  }
+
+  std::copy(left.begin(), left.end(), std::back_inserter(leftBucketed));
+  std::copy(right.begin(), right.end(), std::back_inserter(rightBucketed));
+  moveFloatColumnsToEnd(floatColumnIndices, width, leftBucketed);
+  moveFloatColumnsToEnd(floatColumnIndices, width, rightBucketed);
+  std::sort(leftBucketed.begin(), leftBucketed.end());
+  std::sort(rightBucketed.begin(), rightBucketed.end());
+
+  if (floatColumnIndices.size() <= 1) {
+    return true;
+  }
+
+  // Check that every bucket grouped by non-floating-point columns has only
+  // one row.
+  width = width - floatColumnIndices.size();
+  bool leftOversized = hasMultiRowGroups(leftBucketed, width);
+  bool rightOversized = hasMultiRowGroups(rightBucketed, width);
+
+  return !leftOversized || !rightOversized;
```

Comment:
Should this be && ?

Reply:
If only one of the result sets satisfies the requirement of every group contains a single row, we could still use the sort-merge algorithm with epsilon comparator, because the extra rows in the group in the other result set would be mismatch anyway. 

E.g., expected result is [(1, 0.1, 1.1), (2, 2.2)] and actual result is [(1, 0.09999, 1.09999), (1, 0.10001, 1.100001)]. Sort-merge still works.